pkg/settings/setting.go

```diff
@@ -47,22 +51,66 @@ type NonMaskedSetting interface {
 	// Encoded returns the encoded representation of the current value of the
 	// setting.
 	Encoded(sv *Values) string
+
 	// EncodedDefault returns the encoded representation of the default value of
 	// the setting.
 	EncodedDefault() string
+
 	// SetOnChange installs a callback to be called when a setting's value
 	// changes. `fn` should avoid doing long-running or blocking work as it is
 	// called on the goroutine which handles all settings updates.
 	SetOnChange(sv *Values, fn func(ctx context.Context))
+
 	// ErrorHint returns a hint message to be displayed to the user when there's
 	// an error.
 	ErrorHint() (bool, string)
 }
 
-// Visibility describes how a user should feel confident that
-// they can customize the setting.  See the constant definitions below
-// for details.
-type Visibility int
+// Class describes the scope of a setting in multi-tenant scenarios. While all
+// settings can be used on the system tenant, the classes restrict use on
+// non-system tenants.
+//
+// Settings can only be registered via the Class, e.g.
+// SystemOnly.RegisterIntSetting().
+//
+// Guidelines for choosing a class:
+//  - Make sure to read the descriptions below carefully to understand the
+//    differences in semantics.
+//
+//  - If the setting controls a user-visible aspect of SQL, it should be a
+//    TenantWritable setting.
+//
+//  - Control settings relevant to tenant-specific internal implementation
+//    should be TenantReadOnly.
+//
+//  - When in doubt, the first choice to consider should be TenantReadOnly.
+//
+//  - SystemOnly should be used with caution: even internal tenant code is
+//    disallowed from using these settings at all.
+type Class int8
+
+const (
+	// SystemOnly settings are associated with single-tenant clusters and host
+	// clusters. Settings with this class do not exist on non-system tenants and
+	// can only be used by the system tenant.
+	SystemOnly Class = iota
+
+	// TenantReadOnly settings are visible to non-system tenants but cannot be
+	// modified by the tenant. Values for these settings are set from the system
+	// tenant and propagated from the host cluster.
+	TenantReadOnly
+
+	// TenantWritable settings are visible to and can be modified by non-system
+	// tenants. The system can still override these settings; the overrides are
+	// propagated from the host cluster.
+	TenantWritable
```

Comment:
just curios: is the purpose of this API change to ensure that each cluster setting is annotated with appropriate scope and that there is no way to forget to do so, like it'd be possible if "WithSystemOnly" was augmented with `WithTeanantReadonly`?

Have we considered linters perhaps to ensure that With* always applied when registering settings to enforce that (as an alternative to this change)?

Is it possible that at some point we would want to express more complex class of settings (perhaps TennantWritable that's not a System visible or overridable)?  Seems like the "With*" approach might be more flexible...